### PR TITLE
Replace Octree by DynamicBVH in cull code

### PR DESCRIPTION
**Pull request by reduz**
-Much greater pairing/unpairing performance
-For now, using it for culling too, but this will change in a couple of days.
-Added a paged allocator, to efficiently alloc/free some types of objects.

### CODE REVIEW

**Comment by Calinou** [comment]
Does this PR fix the issue where hidden nodes are still counted in the culling code, therefore decreasing performance?

**Comment by reduz** [comment]
> Does this PR fix the issue where hidden nodes are still counted in the culling code, therefore decreasing performance?

I think so, hidden nodes are no longer part of the culling nor inserted into the bvh.

**Review by clayjohn — CHANGES_REQUESTED** [review]
I only reviewed for obvious typos and errors.

File: servers/rendering_server.cpp [hunk 1]
```diff
@@ -2359,6 +2359,9 @@ RenderingServer::RenderingServer() {
 	ProjectSettings::get_singleton()->set_custom_property_info("rendering/volumetric_fog/directional_shadow_shrink", PropertyInfo(Variant::INT, "rendering/volumetric_fog/directional_shadow_shrink", PROPERTY_HINT_RANGE, "32,2048,1"));
 	GLOBAL_DEF("rendering/volumetric_fog/positional_shadow_shrink", 512);
 	ProjectSettings::get_singleton()->set_custom_property_info("rendering/volumetric_fog/positional_shadow_shrink", PropertyInfo(Variant::INT, "rendering/volumetric_fog/positional_shadow_shrink", PROPERTY_HINT_RANGE, "32,2048,1"));
+
+	GLOBAL_DEF("rendering/spatial_indexer/update_iterations_per_frame", 10);
+	ProjectSettings::get_singleton()->set_custom_property_info("rendering/dynamic_bvh/update_iterations_per_frame", PropertyInfo(Variant::INT, "rendering/dynamic_bvh/update_iterations_per_frame", PROPERTY_HINT_RANGE, "0,1024,1"));
```
[the remark this answers is not in the document]
Comment (reduz) [reply servers/rendering_server.cpp:2364]: ah my mistake thanks

**Review by clayjohn — APPROVED** [review]
Approved so we can get this merged.

**Comment by lawnjelly** [comment]
I don't have IRC logs so I don't know if any changes were since but this was for the version from 14 hours ago:

I added some basic debug print of the bvh tree:
https://pastebin.com/BcWjD7zh
This is adding 8 random boxes...
```
<lawnjelly> So I'm not sure that the optimize is doing anything
<lawnjelly> and the tree is staying in worst case balance I think.
<lawnjelly> I'll have another look tomorrow as it is bedtime here.
<lawnjelly> And I'll do a PR to your branch maybe with the debug output code (or you can write some is easy).
<lawnjelly> N is node and L is leaf node by the way.
<lawnjelly> When balanced well it should have more nodes at the top and leafs towards the bottom.
<lawnjelly> Or maybe the optimize routine in bullet isn't able to do that kind of balancing.
<lawnjelly> if it can't the Box2D or randy gaul's can I think.
```
I'll take another look now and see if I can figure out why it is not balancing.

**Comment by lawnjelly** [comment]
Further debugging shows the reason why balance isn't working:
```
optimize_incremental
N [-1, -1, -1, 1, 1.00001, 1.00001]
	L [-1, -1, -1, 1, 1.00001, 1.00001]
	N [-1, -1, -1, 1, 1.00001, 1.00001]
		L [-1, -1, -1, 1, 1.00001, 1.00001]
		N [-1, -1, -1, 1, 1.00001, 1.00001]
			L [-1, -1, -1, 1, 1.00001, 1.00001]
			N [-1, -1, -1, 1, 1.00001, 1.00001]
				L [-1, -1, -1, 1, 1.00001, 1.00001]
				N [-1, -1, -1, 1, 1.00001, 1.00001]
					L [-1, -1, -1, 1, 1.00001, 1.00001]
					N [-1, -1, -1, 1, 1.00001, 1.00001]
						L [-1, -1, -1, 1, 1.00001, 1.00001]
						N [-1, -1, -1, 1, 1.00001, 1.00001]
							L [-1, -1, -1, 1, 1.00001, 1.00001]
							L [-1, -1, -1, 1, 1.00001, 1.00001]
```
This is showing the volumes for each node. The volumes are incorrect. Presumably they are in local space, perhaps before local transform, not in world space.

**Comment by lawnjelly** [comment]
renderer_scene_cull, line 1073 (`_update_instance()`), the aabbs should presumably be transformed_aabbs as follows:
```
			p_instance->indexer_id = p_instance->scenario->indexers[Scenario::INDEXER_GEOMETRY].insert(p_instance->transformed_aabb, p_instance);
		} else {
			p_instance->indexer_id = p_instance->scenario->indexers[Scenario::INDEXER_VOLUMES].insert(p_instance->transformed_aabb, p_instance);
```
The `update` function in the bvh doesn't appear to be being called at all, as far as I can tell so far, only the insert, so that is another thing which we can fix.

**Pull request merged**